README.md
[![MELPA](https://melpa.org/packages/helm-cider-badge.svg)](https://melpa.org/#/helm-cider)
[![License GPL 3][badge-license]](http://www.gnu.org/licenses/gpl-3.0.txt)

# Helm Cider

[Helm](https://github.com/emacs-helm/helm) interface to [CIDER](https://github.com/clojure-emacs/cider).

## Installation

Install via [MELPA](https://melpa.org/#/).

For general information on installing Emacs packages, see the [Emacs Wiki](https://www.emacswiki.org/emacs/InstallingPackages).

## Setup

In your init file, add the following to activate `helm-cider-mode` and use Helm versions of CIDER commands:

```emacs-lisp
(helm-cider-mode 1)
```

Use `M-x helm-cider-mode` to turn the minor mode off, or call

```emacs-lisp
(helm-cider-mode -1)
```

## Features

The following assumes that `helm-cider-mode` is turned on, and that the reader is familiar with [Helm](https://github.com/emacs-helm/helm) functionality.

### Apropos

Calling `cider-apropos` or `cider-apropos-select` brings up a Helm buffer of all symbols across all namespaces (except those excluded using the `helm-cider-apropos-excluded-ns` customizable variable).

Each Helm source is a namespace. To select a namespace, use `C-c n` (the default value of `helm-cider-apropos-ns-key`) and select the desired namespace.

![Apropos](./doc/apropos.png)

### Apropos documentation

Calling `cider-apropos-documentation` or `cider-apropos-documentation-select` brings up a Helm buffer of all symbols across all namespaces with documentation.

Select a namespace as with normal apropos.

If customizable variable `helm-cider-apropos-full-doc` is `t`, show full instead of short documentation. Toggle between full/short documentation with `C-]`.

![Apropos documentation](./doc/apropos-documentation.png)

### Namespaces

Calling `cider-browse-ns` or `cider-browse-ns-all` brings up a Helm buffer of namespaces.

Use `RET` to select a namespace and view symbols in that namespace *without* documentation. Use `S-RET` to select a namespace and view symbols in that namespace *with* documentation.

![Browse namespaces](./doc/browse-ns.png)

### Specs

Calling `helm-cider-spec` brings up a Helm buffer of all spec names across all namespaces.

Each Helm source is a namespace. To select a namespace, use `C-c n` (the default value of `helm-cider-spec-ns-key`) and select the desired namespace.

![Specs](./doc/specs.png)

### Spec namespaces

Calling `helm-cider-spec-ns` brings up a Helm buffer of [spec](https://clojure.org/about/spec) namespaces.

Use `RET` to select a namespace and view spec names in that namespace.

![Spec namespaces](./doc/spec-ns.png)

### REPL history

Calling `helm-cider-repl-history` (bound to `C-c C-l` by default) in a CIDER REPL brings up a Helm buffer of REPL entries.

Use `RET` to select an entry and insert it into the REPL.

Use `f2` to delete an entry from history. Multiple entries can be marked and deleted.

![REPL History](./doc/repl-history.png)

### Clojure cheatsheet

Calling `helm-cider-cheatsheet` brings up a Clojure cheatsheet. This is largely a port of @krisajenkins's [`clojure-cheatsheet`](https://github.com/clojure-emacs/clojure-cheatsheet).

Matches are against both candidates and categories. For example, `arith` will match against the `Primitives : Numbers : Arithmetic` category.

![Clojure cheatsheet](./doc/cheatsheet.png)

## See Also

- CIDER issue [#1541](https://github.com/clojure-emacs/cider/issues/1541), regarding [Ido](https://www.gnu.org/software/emacs/manual/html_mono/ido.html)/Helm support for `apropos`.
- CIDER issue [#1059](https://github.com/clojure-emacs/cider/issues/1059), regarding seeing more completion candidates and grouping them by namespace.
- [helm-clojure](https://github.com/prepor/helm-clojure), an earlier project with similar ideas.

[badge-license]: https://img.shields.io/badge/license-GPLv3-blue.svg


CHANGELOG.md
# Changelog

## master (unreleased)

## 0.5.0 (2022-01-03)

### New features

- [#10](https://github.com/clojure-emacs/helm-cider/pull/10): Port `clojure-cheatsheet`.

### Changed

- [#13](https://github.com/clojure-emacs/helm-cider/pull/13): Remove `cider-compat` require (file was removed from CIDER).
- Require Emacs 26 (follows upstream changes in CIDER).

## 0.4.0 (2018-01-20)

### New features

* 9480e96: Add spec browser


helm-cider-spec.el
;;; helm-cider-spec.el --- Helm interface to CIDER spec browser        -*- lexical-binding: t; -*-

;; Copyright (C) 2016-2018 Tianxiang Xiong

;; Author: Tianxiang Xiong <tianxiang.xiong@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Helm interface to CIDER spec browser

;;; Code:

(require 'cider-client)
(require 'cl-lib)
(require 'helm-cider-util)
(require 'subr-x)


;;;; Customize

(defgroup helm-cider-spec nil
  "Helm interface to CIDER spec browser."
  :prefix "helm-cider-spec-"
  :group 'helm-cider
  :tag "Helm CIDER Spec")

(defcustom helm-cider-spec-ns-compare-fn
  #'helm-cider-spec--ns-compare-fn
  "Function to compare spec namespaces for sorting.

Comparison function takes two arguments, ns-1 and ns-2.  If ns-1
< ns-2, it appears earlier as a Helm source in `helm-cider-spec'
and candidate in `helm-cider-spec-ns'."
  :group 'helm-cider-spec
  :type 'function)

(defcustom helm-cider-spec-ns-key "C-c n"
  "String representation of key sequence for executing
`helm-cider-spec-ns'.

This is intended to be added to the keymap for
`helm-cider-spec-symbol'."
  :group 'helm-cider-spec
  :type 'key-sequence)

(defcustom helm-cider-spec-follow nil
  "If non-nil, turn on `helm-follow-mode' for CIDER specs."
  :group 'helm-cider-spec
  :type 'boolean)

(defcustom helm-cider-spec-actions
  (helm-make-actions
   "View spec" #'cider-browse-spec--browse)
  "Actions for Helm CIDER specs."
  :group 'helm-cider-spec
  :type '(alist :key-type string :value-type function))

(defcustom helm-cider-spec-ns-actions
  (helm-make-actions
   "Search in namespace" #'helm-cider-spec-symbol)
  "Actions for Helm CIDER spec namespaces."
  :group 'helm-cider-spec
  :type '(alist :key-type string :value-type function))


;;;; Spec

(defun helm-cider-spec--hashtable (spec-names)
  "Build a hash table from list of SPECS-NAMES.

Keys are kw namespaces and values are lists of names."
  (let ((ht (make-hash-table :test #'equal)))
    (dolist (name spec-names)
      (let ((ns (helm-cider--symbol-ns name)))
        (puthash ns (cons name (gethash ns ht)) ht)))
    (cl-loop
       for ns being the hash-keys of ht
       using (hash-value names)
       do (setf (gethash ns ht) (sort names #'string<))
       finally (return ht))))

(defun helm-cider-spec--candidate (spec-name)
  "Create a Helm CIDER spec candidate.

SPEC-NAME is a spec keyword string."
  (cons (propertize (helm-cider--symbol-name spec-name)
                    'face 'clojure-keyword-face)
        spec-name))

(defvar helm-cider-spec--map
  (let ((keymap (copy-keymap helm-map)))
    ;; Select ns
    (define-key keymap (kbd helm-cider-spec-ns-key)
      (lambda ()
        (interactive)
        (helm-exit-and-execute-action (lambda (candidate)
                                        (helm-cider-spec-ns candidate)))))
    keymap)
  "Keymap for use with `helm-cider-spec'.")

(defun helm-cider-spec--persistent-action (candidate)
  "Persistent action for Helm CIDER apropos."
  (if (and (helm-attr 'spec-lookup-p)
           (string= candidate (helm-attr 'current-candidate)))
      (progn
        (kill-buffer cider-browse-spec-buffer)
        (helm-attrset 'spec-lookup-p nil))
    (cider-browse-spec--browse candidate)
    (helm-attrset 'spec-lookup-p t))
  (helm-attrset 'current-candidate candidate))

(defun helm-cider-spec--source (ns spec-names &optional follow)
  "Helm source for specs in namespace NS.

SPEC-NAMES is a list of spec keywords strings."
  (helm-build-sync-source ns
    :action helm-cider-spec-actions
    :candidates (mapcar #'helm-cider-spec--candidate spec-names)
    :follow (when follow 1)
    :keymap helm-cider-spec--map
    :nomark t
    :persistent-action #'helm-cider-spec--persistent-action
    :persistent-help "View spec"
    :volatile t))

(defun helm-cider-spec--ns-compare-fn (ns-1 ns-2)
  "Function to compare spec namespaces NS-1 and NS-2.

Namespaces are compared as symbols, without keywords' leading
colons."
  (let ((ns-1 (replace-regexp-in-string "^:" "" ns-1))
        (ns-2 (replace-regexp-in-string "^:" "" ns-2)))
    (string< ns-1 ns-2)))

(defun helm-cider-spec--sources (&optional spec-names)
  "A list of Helm sources specs.

Each source is the set specs in a namespace.

Optional argument SPEC-NAMES is a list of spec keyword strings.
If not supplied, it is retrieved with
`cider-sync-request:spec-list'."
  (cl-loop
     with ht = (helm-cider-spec--hashtable (or spec-names
                                               (cider-sync-request:spec-list "")))
     for ns being the hash-keys in ht using (hash-value names)
     collect (helm-cider-spec--source ns names helm-cider-spec-follow)
     into sources
     finally (return (cl-sort sources helm-cider-spec-ns-compare-fn
                              :key (lambda (source) (assoc-default 'name source))))))

(defun helm-cider-spec--all-ns ()
  "Return a list of all spec namespace strings.

E.g. '(\":ring.async.handler\", \":ring.core\", ...)."
  (let ((nss (thread-last (cider-sync-request:spec-list "")
               (mapcar #'helm-cider--symbol-ns))))
    (cl-delete-duplicates nss :test #'equal)))

(defun helm-cider-spec--propertize-ns (ns)
  "Propertize a spec namespace NS.

I.e. how `:foo' would be propertized in `:foo/bar' in
`clojure-mode'.

It is also possible for a spec ns to be a regular symbol
instead of keyword, as with `clojure.core'."
  (if (string-prefix-p ":" ns)
      (concat (propertize (substring ns 0 1)
                          'face 'clojure-keyword-face)
              (propertize (substring ns 1)
                          'face 'font-lock-type-face))
    (propertize ns 'face 'font-lock-type-face)))

(defun helm-cider-spec--ns-source (&optional follow)
  "Helm source of namespaces.

Namespaces in EXCLUDED-NS are excluded.  If not supplied,
`helm-cider-apropos-excluded-ns' is used.

If FOLLOW is true, use function `helm-follow-mode' for source."
  (helm-build-sync-source "Clojure Spec Namespaces"
    :action helm-cider-spec-ns-actions
    :candidates (cl-loop
                   for ns in (helm-cider-spec--all-ns)
                   collect (helm-cider-spec--propertize-ns ns) into all
                   finally (return (sort all helm-cider-spec-ns-compare-fn)))
    :follow (when follow 1)
    :nomark t
    :persistent-action #'ignore
    :volatile t))

(defun helm-cider-spec--resolve-name (&optional ns name)
  "Try to get correct values for NS and NAME.

NS is a spec keyword ns, e.g. \":ring.core\". NAME is a spec
keyword name, e.g. \"error\"."
  (let* ((name (or name (unless ns (cider-symbol-at-point))))
         (qualified (when name
                      (or (cider-namespace-qualified-p name)
                          (string-match-p "^::" name))))
         (ns (cond (qualified (if (string-match-p "^::" name)
                                  (cider-current-ns)
                                (helm-cider--symbol-ns name)))
                   (t ns))))
    (list ns (if qualified
                 (helm-cider--symbol-name name)
               name))))


;;;; API

;;;###autoload
(defun helm-cider-spec-symbol (&optional ns name)
  "Choose Clojure specs across namespaces.

Each Helm source is a Clojure namespace (ns), and candidates are
spec keywords in the namespace.

If both NS and NAME are supplied, puts selection line on
first NAME of NS.

If NS is supplied, puts the selection line on the first
candidate of source with name NS.

If NAME is supplied, puts the selection line on the
first candidate matching NAME.

Set `helm-cider-spec-follow' to non-nil to turn on function
`helm-follow-mode' for all sources.  This is useful for quickly
viewing specs."
  (interactive)
  (cider-ensure-connected)
  (cl-multiple-value-bind (ns name) (helm-cider-spec--resolve-name ns name)
    (let ((name (when name
                  (helm-cider--regexp-symbol name))))
      (with-helm-after-update-hook
        (with-helm-buffer
          (let ((helm--force-updating-p t))
            (if name
                (helm-preselect name (helm-cider--source-by-name ns))
              (helm-goto-source (or ns ""))
              (when ns
                (helm-next-line)))
            (recenter 1))))
      (helm :buffer "*Helm Clojure Specs*"
            :candidate-number-limit 9999
            :sources (helm-cider-spec--sources)))))

;;;###autoload
(defun helm-cider-spec-ns (&optional kw-ns-or-qualified-name)
  "Choose spec namespace to call `helm-cider-browse-spec' on.

KW-NS-OR-QUALIFIED-NAME is a spec keyword namespace
 (e.g. \":ring.core\") or a qualified keyword
name (e.g. \":ring.core/error\").  If supplied, it is used as the
default selection."
  (interactive)
  (cider-ensure-connected)
  (helm :buffer "*Helm Clojure Spec Namespaces*"
        :candidate-number-limit 9999
        :preselect (thread-first (or kw-ns-or-qualified-name "")
                     helm-cider--symbol-ns
                     helm-cider--regexp-symbol)
        :sources (helm-cider-spec--ns-source)))

;;;###autoload
(defun helm-cider-spec (&optional arg)
  "Helm interface to CIDER specs.

If ARG is raw prefix argument \\[universal-argument]
\\[universal-argument], choose namespace before symbol."
  (interactive "P")
  (cond ((equal arg '(4)) (helm-cider-spec-ns))
        (t (helm-cider-spec-symbol))))



(provide 'helm-cider-spec)

;;; helm-cider-spec.el ends here


helm-cider.el
;;; helm-cider.el --- Helm interface to CIDER        -*- lexical-binding: t; -*-

;; Copyright (C) 2016-2022 Tianxiang Xiong

;; Author: Tianxiang Xiong <tianxiang.xiong@gmail.com>
;; Package-Requires: ((emacs "26") (cider "1.0") (helm-core "2.8"))
;; Keywords: cider, clojure, helm, languages
;; URL: https://github.com/clojure-emacs/helm-cider
;; Version: 0.4.0

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Helm interface to CIDER.

;; For more about Helm, see: https://github.com/emacs-helm/helm
;; For more about CIDER, see: https://github.com/clojure-emacs/cider

;;; Code:

(require 'cider)
(require 'cl-lib)
(require 'helm)
(require 'helm-cider-cheatsheet)
(require 'helm-cider-repl)
(require 'helm-cider-spec)
(require 'helm-cider-util)
(require 'helm-lib)
(require 'helm-multi-match)
(require 'helm-source)


;;;; Customize
(defgroup helm-cider nil
  "Helm interface to CIDER."
  :prefix "helm-cider-"
  :group 'cider
  :tag "Helm CIDER")

;;;;; Apropos

(defgroup helm-cider-apropos nil
  "Helm CIDER apropos"
  :prefix "helm-cider-apropos-"
  :group 'helm-cider
  :tag "Helm CIDER Apropos")

(defcustom helm-cider-apropos-excluded-ns '("cider.*")
  "List of namespaces to exclude from apropos.

Namespace globs (e.g. \"cider.*\" for all CIDER-specific
namespaces) are accepted.

By default, CIDER-specific namespaces (those used by CIDER
itself, e.g. \"cider.nrepl.middleware.apropos\") are excluded."
  :group 'helm-cider-apropos
  :type '(repeat string))

(defcustom helm-cider-apropos-follow nil
  "If true, enable function `helm-follow-mode' for apropos sources."
  :group 'helm-cider-apropos
  :type 'boolean)

(defcustom helm-cider-apropos-ns-actions
  (helm-make-actions
   "Search in namespace" #'helm-cider-apropos-symbol
   "Find definition" #'helm-cider--find-ns
   "CiderDoc" (wrap-helm-cider-action cider-doc-lookup)
   "Search in namespace with docs" #'helm-cider-apropos-symbol-doc
   "Set REPL namespace" #'cider-repl-set-ns)
  "Actions for Helm apropos namespaces."
  :group 'helm-cider-apropos
  :type '(alist :key-type string :value-type function))

(defcustom helm-cider-apropos-actions
  helm-cider--doc-actions
  "Actions for Helm apropos symbols."
  :group 'helm-cider-apropos
  :type '(alist :key-type string :value-type function))

(defcustom helm-cider-apropos-full-doc nil
  "If true, use full documentation in `helm-cider-apropos-symbol-doc'."
  :group 'helm-cider-apropos
  :type 'boolean)

(defcustom helm-cider-apropos-ns-key "C-c n"
  "String representation of key sequence for executing
`helm-cider-apropos-ns'.

This is intended to be added to the keymap for
`helm-cider-apropos'."
  :group 'helm-cider-apropos
  :type 'key-sequence)

(defcustom helm-cider-apropos-symbol-doc-key "S-<return>"
  "String representation of key sequence of executing
`helm-cider-apropos-symbol-doc'.

This is intended to be added to the keymap for
`helm-cider-apropos-ns.'"
  :group 'helm-cider-apropos
  :type 'key-sequence)


;;;; Apropos

(defun helm-cider--excluded-ns-p (ns &optional excluded-ns)
  "Return true when namespace NS matches one of EXCLUDED-NS.

EXCLUDED-NS is a list of namespaces (e.g. \"clojure.core\")
and/or namespace globs (e.g. \"cider.*\"). If not supplied,
`helm-cider-apropos-excluded-ns' is used.

NS matches a string equal to itself, or a string ending in \"*\"
that is a prefix of NS, excluding the \"*\"."
  (catch 'excluded
    (dolist (ex (or excluded-ns
                    helm-cider-apropos-excluded-ns))
      (when (or (and (string-suffix-p "*" ex)
                     (string-prefix-p (substring ex 0 (1- (length ex))) ns))
                (string= ns ex))
        (throw 'excluded t)))))

(defun helm-cider--apropos-dicts (&optional excluded-ns full-doc)
  "List of apropos results (nREPL dicts).

Optional argument EXCLUDED-NS is a list of ns for which symbols
are excluded.  If not supplied, `helm-cider-apropos-excluded-ns'
is used.

Optional argument FULL-DOC, if t, retrieves full documentation."
  (cl-loop
     with excluded-ns = (or excluded-ns helm-cider-apropos-excluded-ns)
     for dict in (cider-sync-request:apropos "" nil full-doc)
     unless (helm-cider--excluded-ns-p (helm-cider--symbol-ns (nrepl-dict-get dict "name"))
                                       excluded-ns)
     do (unless (nrepl-dict-contains dict "ns")
          (nrepl-dict-put dict "ns" "clojure.core"))
     collect dict))

(defun helm-cider--apropos-hashtable (dicts)
  "Build a hash table from apropos results (DICTS).

Keys are namespaces and values are lists of results (nREPL
dict objects)."
  (let ((ht (make-hash-table :test 'equal)))
    (dolist (dict dicts)
      (let ((ns (helm-cider--symbol-ns (nrepl-dict-get dict "name"))))
        (puthash ns (cons dict (gethash ns ht)) ht)))
    ht))

(defun helm-cider--apropos-candidate (dict)
  "Create a Helm apropos candidate.

DICT is an nREPL dict."
  (nrepl-dbind-response dict (name type)
    (cons (propertize (helm-cider--symbol-name name)
                      'face (helm-cider--symbol-face type))
          name)))

(defun helm-cider--apropos-doc-candidate (dict &optional full-doc)
  "Create a Helm apropos doc candidate.

DICT is an nREPL dict.

Optional argument FULL-DOC, when t, indicates that full
documentation is used."
  (nrepl-dbind-response dict (name type doc)
    (with-temp-buffer
      ;; Name
      (insert (propertize name 'face (helm-cider--symbol-face type)))
      (insert "\n")
      ;; Doc
      (let ((beg (point)))
        (insert doc "\n")
        ;; Short doc is not already filled; full doc is
        (when (not full-doc)
          (fill-region beg (point-max))))
      ;; Candidate
      (cons (buffer-string) name))))

(defvar helm-cider--apropos-map
  (let ((keymap (copy-keymap helm-map)))
    ;; Select namespace
    (define-key keymap (kbd helm-cider-apropos-ns-key)
      (lambda ()
        (interactive)
        (helm-exit-and-execute-action (lambda (candidate)
                                        (helm-cider-apropos-ns candidate)))))
    ;; Apropos with docs
    (dolist (key (cl-mapcan (lambda (command)
                              (where-is-internal command cider-mode-map))
                            '(cider-apropos-documentation
                              cider-apropos-documentation-select)))
      (define-key keymap key
        (lambda ()
          (interactive)
          (helm-exit-and-execute-action
           (lambda (candidate)
             (helm-cider-apropos-symbol-doc (helm-cider--symbol-ns candidate)
                                            (helm-cider--symbol-name candidate)))))))
    ;; Apropos
    (dolist (key (cl-mapcan (lambda (command)
                              (where-is-internal command cider-mode-map))
                            '(cider-apropos
                              cider-apropos-select)))
      (define-key keymap key
        (lambda ()
          (interactive)
          (helm-exit-and-execute-action
           (lambda (candidate)
             (helm-cider-apropos-symbol (helm-cider--symbol-ns candidate)
                                        (helm-cider--symbol-name candidate)))))))
    keymap)
  "Keymap for use with `helm-cider-apropos-symbol'.")

(defvar helm-cider--apropos-doc-map
  (let ((keymap (copy-keymap helm-cider--apropos-map)))
    ;; Toggle full documentation
    (define-key keymap (kbd "C-]")
      (lambda ()
        (interactive)
        (helm-exit-and-execute-action
         (lambda (candidate)
           (customize-set-value 'helm-cider-apropos-full-doc
                                (not helm-cider-apropos-full-doc))
           (helm-cider-apropos-symbol-doc (helm-cider--symbol-ns candidate)
                                          (helm-cider--symbol-name candidate))))))
    keymap)
  "Keymap for use with `helm-cider-apropos-symbol-doc'.")

(defun helm-cider--apropos-source (ns &optional dicts doc full-doc follow)
  "Helm source for namespace NS (e.g. \"clojure.core\").

DICTS is a list of apropos results (nREPL dicts) for
NS. If not supplied, it is obtained with
`cider-sync-request:apropos'.

If DOC is true, include symbol documentation in candidates.

If FULL-DOC is true, full documentation is used; candidates may
need to be treated differently by
`helm-cider--apropos-doc-candidate'.

If FOLLOW is true, use function `helm-follow-mode' for source."
  (helm-build-sync-source ns
    :action helm-cider-apropos-actions
    :candidates (thread-first (if doc
                                  (lambda (dict)
                                    (helm-cider--apropos-doc-candidate dict full-doc))
                                #'helm-cider--apropos-candidate)
                  (mapcar (if dicts
                              (copy-sequence dicts)
                            (cider-sync-request:apropos "" ns doc)))
                  (cl-sort #'string< :key (lambda (dict) (helm-cider--symbol-name (cdr dict)))))
    :follow (when follow 1)
    :keymap (if doc helm-cider--apropos-doc-map helm-cider--apropos-map)
    :multiline doc
    :nomark t
    :persistent-action #'helm-cider--doc-lookup-persistent-action
    :persistent-help "Look up documentation"
    :volatile t))

(defun helm-cider--apropos-sources (&optional excluded-ns doc full-doc)
  "A list of Helm sources for apropos.

Each source is the set of symbols in a namespace.  Namespaces in
EXCLUDED-NS are excluded.  If not supplied,
`helm-cider-apropos-excluded-ns' is used.

If DOC is true, include symbol documentation in candidates.

If FULL-DOC is true, include full instead of short documentation."
  (cl-loop
     with ht = (helm-cider--apropos-hashtable
                (helm-cider--apropos-dicts excluded-ns full-doc))
     for ns being the hash-keys in ht using (hash-value dicts)
     collect (helm-cider--apropos-source ns dicts doc full-doc helm-cider-apropos-follow)
     into sources
     finally (return (cl-sort sources #'string< :key (lambda (source) (assoc-default 'name source))))))

(defvar helm-cider--apropos-ns-map
  (let ((keymap (copy-keymap helm-map)))
    (define-key keymap (kbd helm-cider-apropos-symbol-doc-key)
      (lambda ()
        (interactive)
        (helm-exit-and-execute-action (lambda (candidate)
                                        (helm-cider-apropos-symbol-doc candidate)))))
    keymap)
  "Return a keymap for use with `helm-cider-apropos-ns'.")

(defun helm-cider--apropos-ns-source (&optional excluded-ns follow)
  "Helm source of namespaces.

Namespaces in EXCLUDED-NS are excluded.  If not supplied,
`helm-cider-apropos-excluded-ns' is used.

If FOLLOW is true, use function `helm-follow-mode' for source."
  (helm-build-sync-source "Clojure Namespaces"
    :action helm-cider-apropos-ns-actions
    :candidates (cl-loop for ns in (cider-sync-request:ns-list)
                   unless (helm-cider--excluded-ns-p ns excluded-ns)
                   collect (cider-propertize ns 'ns) into all
                   finally (return (sort all #'string<)))
    :follow (when follow 1)
    :keymap helm-cider--apropos-ns-map
    :nomark t
    :persistent-action #'helm-cider--doc-lookup-persistent-action
    :persistent-help "Look up documentation"
    :volatile t))

(defun helm-cider--resolve-symbol (&optional ns symbol)
  "Try to get correct values for NS and SYMBOL."
  (let* ((symbol (or symbol (unless ns (cider-symbol-at-point))))
         (qualified (when symbol (cider-namespace-qualified-p symbol)))
         (ns (cond (qualified (nrepl-dict-get (cider-var-info symbol t) "ns"))
                   ((and ns symbol)
                    (nrepl-dict-get (cider-var-info (concat ns "/" symbol) t) "ns"))
                   (symbol (nrepl-dict-get (cider-var-info symbol t) "ns"))
                   (t ns))))
    (list ns (if qualified
                 (helm-cider--symbol-name symbol)
               symbol))))


;;;; API

;;;;; Apropos

;;;###autoload
(defun helm-cider-apropos-symbol (&optional ns symbol doc full-doc)
  "Choose Clojure symbols across namespaces.

Each Helm source is a Clojure namespace (ns), and candidates are
symbols in the namespace.

If both NS and SYMBOL are supplied, puts selection line on
first SYMBOL of NS.

If NS is supplied, puts the selection line on the first
candidate of source with name NS.

If SYMBOL is supplied, puts the selection line on the
first candidate matching SYMBOL.

If neither NS nor SYMBOL is supplied, tries to put the
selection line on candidate matching symbol at point.

If DOC is true, include symbol documentation in candidate.

If FULL-DOC is true, include full instead of short documentation.

Set `helm-cider-apropos-follow' to true to turn on function
`helm-follow-mode' for all sources.  This is useful for quickly
browsing documentation."
  (interactive)
  (cider-ensure-connected)
  (cl-multiple-value-bind (ns symbol) (helm-cider--resolve-symbol ns symbol)
    (let ((symbol (cond ((and symbol doc) (regexp-quote (concat ns "/" symbol)))
                        (symbol (helm-cider--regexp-symbol symbol)))))
      (with-helm-after-update-hook
        (with-helm-buffer
          (let ((helm--force-updating-p t))
            (if symbol
                (helm-preselect symbol (helm-cider--source-by-name ns))
              (helm-goto-source (or ns ""))
              (when ns
                (helm-next-line)))
            (recenter 1))))
      (helm :buffer "*Helm Clojure Symbols*"
            :candidate-number-limit 9999
            :sources (helm-cider--apropos-sources nil doc full-doc)))))

;;;###autoload
(defun helm-cider-apropos-symbol-doc (&optional ns symbol)
  "Choose Clojure SYMBOLs, with docs, across namespaces.

Optional arguments NS and SYMBOL are as in
`helm-cider-apropos-symbol'."
  (interactive)
  (helm-cider-apropos-symbol ns symbol t helm-cider-apropos-full-doc))

;;;###autoload
(defun helm-cider-apropos-ns (&optional ns-or-qualified-name)
  "Choose Clojure namespace to call Helm CIDER apropos on.

NS-OR-QUALIFIED-NAME is a Clojure
namespace (e.g. \"clojure.core\") or a qualified symbol
name (e.g. \"clojure.core/reduce\").  If supplied, it is used as
the default selection."
  (interactive)
  (cider-ensure-connected)
  (helm :buffer "*Helm Clojure Namespaces*"
        :candidate-number-limit 9999
        :preselect (helm-cider--regexp-symbol
                    (helm-cider--symbol-ns (or ns-or-qualified-name "")))
        :sources (helm-cider--apropos-ns-source)))

;;;###autoload
(defun helm-cider-apropos (&optional arg)
  "Helm interface to CIDER apropos.

If ARG is raw prefix argument \\[universal-argument], include
symbol documentation.

If ARG is raw prefix argument \\[universal-argument]
\\[universal-argument], choose namespace before symbol."
  (interactive "P")
  (cond ((equal arg '(16)) (helm-cider-apropos-ns))
        ((equal arg '(4)) (helm-cider-apropos-symbol-doc))
        (t (helm-cider-apropos-symbol))))


;;;; Key bindings and minor mode

(defcustom helm-cider-overrides
  '((cider-apropos . helm-cider-apropos)
    (cider-apropos-select . helm-cider-apropos)
    (cider-apropos-documentation . helm-cider-apropos-symbol-doc)
    (cider-apropos-documentation-select . helm-cider-apropos-symbol-doc)
    (cider-browse-ns . helm-cider-apropos-ns)
    (cider-browse-ns-all . helm-cider-apropos-ns)
    (cider-browse-spec-all . helm-cider-spec))
  "Alist of CIDER functions and Helm versions replacing them."
  :group 'helm-cider
  :type '(alist :key-type symbol :value-type symbol))

(defun helm-cider--override ()
  "Override CIDER functions with Helm versions.

The old and new functions are those specified in
`helm-cider-overrides'."
  (dolist (pair helm-cider-overrides)
    (let ((symbol (car pair))
          (newfun (symbol-function (cdr pair))))
      (unless (advice-member-p newfun symbol)
        (advice-add symbol :override newfun)))))

(defun helm-cider--revert ()
  "Revert to original CIDER functions."
  (dolist (pair helm-cider-overrides)
    (advice-remove (car pair) (symbol-function (cdr pair)))))

;;;###autoload
(define-minor-mode helm-cider-mode "Use Helm for CIDER."
  :global t
  :require 'helm-cider
  (if helm-cider-mode
      (progn
        (helm-cider--override)
        (define-key cider-repl-mode-map (kbd "C-c C-l") #'helm-cider-repl-history))
    (helm-cider--revert)
    (define-key cider-repl-mode-map (kbd "C-c C-l") nil)))


(provide 'helm-cider)

;;; helm-cider.el ends here
